# CS 237: Homework 2 Programming Exercises

In [1]:
# Jupyter notebook specific 
from IPython.display import Image
from IPython.core.display import HTML 
from IPython.display import display_html
from IPython.display import display
from IPython.display import Math
from IPython.display import Latex
from IPython.display import HTML

import numpy as np                # arrays and functions which operate on array
import matplotlib.pyplot as plt   # normal plotting

from numpy.random import seed, randint, uniform
from collections import Counter

In the following exercises, we will use simulation to estimate the probability of certain events. Write the code to perform the simulation and print the result. Use 10,000 trials in each case.

## Exercise 1: Simulation of the Chevalier's bets

In this exercise, consider the Chevalier de Mere's dice bets from lecture. Use simulation to estimate the probabilities of winning each bet, i.e., the probabilities of the following events:

- The event that a 6 turns up when rolling a die 4 times
- The event that a pair of 6s turns up when rolling two dice 24 times

Print the estimated probabilities. Are your estimates close to the probabilities we calculated in class?

In [2]:
def roll_die_4(n_trials = 10000, seed=-1):
    """
    runs 10,000 trials and records the successes of rolling a die 4 times and getting a 6
    returns the probability calculated as #of successes/#of trials
    """
    
    num_rolls = 1
    success=np.zeros(n_trials)
    for i in range(n_trials):
        while num_rolls<=4:
            die_roll = randint(1, 7) #fair die roll
            if die_roll == 6:
                #stores a successful experiment as 1
                success[i]=1
                break
            num_rolls += 1
        num_rolls=1 
    #returns the probability of getting a success over 10000 trials
    return sum(success)/n_trials

def roll_die_24(n_trials = 10000, seed=-1):
    """
    runs 10,000 trials and records the successes of rolling a pair of dice 24 times and 
    getting a pair of 6s. returns the probability calculated as #of successes/#of trials
    """
    
    num_rolls = 1
    success=np.zeros(n_trials)
    for i in range(n_trials):
        while num_rolls<=24:
            die1 = randint(1, 7)
            die2 = randint(1, 7)
            if die1 == 6 and die2 == 6:
                success[i]=1
                break
            num_rolls += 1
        num_rolls=1 
    return sum(success)/n_trials

print("Probability of getting a 6 after rolling a die 4 times after 10,000 trials is " + str(roll_die_4()))
print()
print("Probability of getting a pair of 6s after rolling dice 24 times after 10,000 trials is " + str(roll_die_24()))


Probability of getting a 6 after rolling a die 4 times after 10,000 trials is 0.5228

Probability of getting a pair of 6s after rolling dice 24 times after 10,000 trials is 0.4904


**Answer:**
    *After running an experiment for the event that a 6 turns up when rolling a die 4 times I got a 6 with probability 0.5228 which is very close to the calculated value of 0.5177.*
    *After running an experiment for the event that a pair of 6s turns up when rolling two dice 24 times I got a pair of 6s with probability 0.4904 which is very close to the calculated value of 0.4914*

## Exercise 2: Sums of 3 dice

In this exercise, we consider another dice betting experiment that puzzled gamblers in the 1600s. Consider the experiment of rolling three dice. We are interested in the following two events: 

- A = "The sum of numbers obtained is 9"
- B = "The sum of the numbers obtained is 10"

Use simulation to estimate the probabilities of the two events. Print the estimated probabilities. According to your simulation, which event is more likely?

In [3]:
def roll_and_sum_9(n_trials = 10000, seed=-1):
    """
    records the number of successes of rolling 3 dice and getting sum of 9
    returns the probabilty of success over 10,000 trials
    """
    success=np.zeros(n_trials)
    dice_sum = 0
    for trials in range(n_trials):
        #simulates rolling 3 dice by creating an array of 3 random int and summing 
        arr = randint(1,7,(3))   
        dice_sum = np.sum(arr)
        if dice_sum == 9:
            success[trials]=1
        dice_sum= 0
    return sum(success)/n_trials

def roll_and_sum_10(n_trials = 10000, seed=-1):
    """
    records the number of successes of rolling 3 dice and getting sum of 10
    returns the probabilty of success over 10,000 trials
    """
    success=np.zeros(n_trials)
    dice_sum = 0
    for trials in range(n_trials):
        #simulates rolling 3 dice by creating an array of 3 random int and summing 
        arr = randint(1,7,(3))   
        dice_sum = np.sum(arr)
        if dice_sum == 10:
            success[trials]=1
        dice_sum= 0
    return sum(success)/n_trials

print("Probability of getting a sum of 9 after rolling 3 dice after 10,000 trials is " +str(roll_and_sum_9()))
print()
print("Probability of getting a sum of 10 after rolling 3 dice after 10,000 trials is " +str(roll_and_sum_10()))

Probability of getting a sum of 9 after rolling 3 dice after 10,000 trials is 0.1168

Probability of getting a sum of 10 after rolling 3 dice after 10,000 trials is 0.1239


The experience of gamblers in the 1600s had been that A is less likely than B. This empirical result puzzled the gamblers for the following reason. They counted the number of triples of integers from 1 to 6 with sum 9 and found it to be the same as the number of triples with sum 10. Can you explain this apparent discrepancy?

**Answer:**
        *After running an experiment for the event that a sum of 9 turns up when rolling 3 dice I got a 9 with probability 0.1168.*
        *After running an experiment for the event that a sum of 10 turns up when rolling 3 dice I got a 10 with probability 0.1239.*
        *The probabilities are very close but event B is a bit more likely because there are 27 ways to make a sum of 10 with 3 integers (with an actual probability of 0.125) while there are 25 ways to make a sum of 9 with 3 integers (and an actual probability of 0.116), hence the slight discrepancy.*


## Exercise 3: Hospitals

In this exercise, we consider another famous study by Tversky and Kahneman. The participants of the study were asked the following question:

"A certain town is served by two hospitals. In the larger hospital about 45 babies are born each day, and in the smaller hospital 15 babies are born each day. As you know, about 50 percent of all babies are boys. However, the exact percentage varies from day to day. Sometimes it may be higher than 50 percent, sometimes lower.

For a period of 1 year, each hospital recorded the days on which more than 60 percent of the babies born were boys. Which hospital do you think recored more such days?

- (a) The larger hospital
- (b) The smaller hospital
- (c) About the same (that is, within 5 percent of each other)
"

For simplicity, we assume that the probability that a baby is a boy is 0.5 and the year has 365 days. Using simulation, determine which is the right answer to the question. Here a single trial simulates each hospital for 1 year.

In [7]:
def hospital_births(n_trials=10000, seed=-1):
    """NOTE takes about 5 min to run 10,000 trials on my machine"""
    
    """
    runs 10,000 trials where in each trial it simulates a year of births for both hospitals
    for each day it records the number of boys born and if at the end of the day it is >60%
    the day is added to the array of years. At the end of each year the number of days with >60%
    boys per hospital is recorded for that year as well as the result to the question for that
    specific year. Finally the arrays containing the results, relevant days for large hospital and 
    relevant days for small hospitals for each year is returned.
    """
    results = []
    large_years=[]
    small_years=[]
    for trial in range(n_trials):
        large_hospital_days=[]
        small_hospital_days=[]
        
        for day in range(365):
            large_birth=[]
            small_birth=[]
            for i in range(45):
                boy1 = randint(0,2)
                if boy1 == 1:
                    large_birth.append(boy1)
            for j in range(15):
                boy2 = randint(0,2)
                if boy2 == 1:
                    small_birth.append(boy2)

            if sum(large_birth)/45>0.6:
                large_hospital_days.append(1)
            
            if sum(small_birth)/15>0.6:
                small_hospital_days.append(1)
                
         
        large_years.append(sum(large_hospital_days))
        small_years.append(sum(small_hospital_days))
        
        if 0.95*sum(large_hospital_days)> sum(small_hospital_days) and 1.05*sum(large_hospital_days)> sum(small_hospital_days) :
            results.append('a')
        elif sum(large_hospital_days)< 0.95*sum(small_hospital_days) and sum(large_hospital_days)< 1.05*sum(small_hospital_days):
            results.append('b')
        else:
            results.append('c')
                                               
    return results, large_years, small_years
                
answer, large, small = hospital_births()

#prints the average number of days with >60% boys for large hospital
print("Average number of days for large hospital after 10,000 trials " + str(sum(large)/len(large)))

#prints the average number of days with >60% boys for large hospital
print("Average number of days for small hospital after 10,000 trials " + str(sum(small)/len(small))) 

#Counts the results per trial and prints them to console
print("Counter of results to the question after 10,000 trials " + str(Counter(answer)))
    

Average number of days for large hospital after 10,000 trials 24.6497
Average number of days for small hospital after 10,000 trials 55.0866
Counter of results to the question after 10,000 trials Counter({'b': 9998, 'c': 2})


 Does the simulation agree with your intuition? (We will not grade this answer, we only want to see how the class distribution looks like.)
 
 **Answer:**
     *According to the results of my simulation the right answer to the question is clearly (b), the smaller hospital recorded more days in which over 60% of the babies born are boys than the larger hospital in virtually every trial, namely 9998 out of 10000. This result agrees with my intuiton since the smaller hospital has only 15 births a day as opposed to 45. The smaller number of births allows for a greater variance in the percentage of boys born. As the number of babies born each day increases we expect the probability to approach 0.5 hence the discrepancy in the percentage of boys born between 2 hospitals.*